### ISSUE TITLE
Update an object in array

### ISSUE THREAD

**Issue by thinklinux** [rethinkdb/docs#935]
Let's say I have table `posts` and every post have an array called `comments` with objects inside to describe the comment like so `comments: [{author: 'Joe', adminComment: true, text: 'Update me'}]`

How would you update the text of the comment? I saw that there is an [issue for that](https://github.com/rethinkdb/rethinkdb/issues/2692) but the solution is not available yet.

[Stack Overflow to the rescue!](http://stackoverflow.com/questions/17133460/how-to-update-embedded-document) There is a good explanation there for the update query. The problem was that it is with hard-coded index for the element of the array. So here is a full example for updating a single comment that worked for me:

```
let post_id = 123;
let comment_id = 23434;

let findCommentIndex =
  r.table("posts")
    .get(post_id)("comments")
    .offsetsOf(
      r.row("id").match(comment_id)
    )
    .nth(0)

let update =
  findCommentIndex.do((index) => {
    return r.table('posts')
      .get(post_id)
      .update((post) => {
        return {
          channels: post('comments').changeAt(index,
            post('comments')
              .nth(index)
              .merge({'text': 'updated!'})
          )
        }
      })
  })
```

This operations should be a lot more easier and I hope someday they will be but right now it would be cool if we have some examples like this one on the docs or cookbook because I'm sure I'm not the only one that lost a lot of time to figure this one out.

**Comment by danielmewes** [comment]
There's also the alternative of using a `map` for finding and modifying the element in a single pass. It shouldn't make a difference for performance, but is sometimes nicer to write:

```js
let update =
  r.table("posts")
   .get(post_id)
   .update({
     comments: r.row("comments").map((comment) => {
         return r.branch(
             comment("id").eq(comment_id),
             {test: 'updated!'},
             {}
         );
       })
   })
```

One problem specific to the code above is that it fetches the post twice, once for `findCommentIndex` and once for `update`. This means that the operation is not atomic, and things might go wrong if there are any concurrent modifications to the post.

**Comment by thinklinux** [comment]
@danielmewes Thanks! More elegant and most important - atomic. Love it!

**Comment by vectorselector** [comment]
Indeed, the docs could be more explicit about updates and what mapping and branching actually do in the context of a Rethinkdb query. (which is far from idiomatic for anyone outside Rethinkland)

Strategic examples would go a LONG way. As it is, I was prepared to ditch Rethinkdb in favor of something inferior simply due to my perceived inability to query and update as required.

If Rethinkdb has such a great query system, DEMONSTRATE IT. Non-trivial examples of deeply nested objects inside arrays etc.

I do apologize for necro-bumping this issue, but if I still find this issue as one of the only sources of info on the world-wide-web in 2018 you can expect such.
